### PR TITLE
Refactor stages

### PR DESCRIPTION
**Pull request by mdellweg**
This is kind of a followup for #32.

Also it is meant as a preparation for https://pulp.plan.io/issues/4296

### CODE REVIEW

**Comment by mdellweg** [comment]
@gmbnomis maybe, you want to take a look, too.

File: pulpcore/plugin/stages/api.py [hunk 1]
```diff
@@ -13,90 +13,144 @@ class Stage:
     To make a stage, inherit from this class and implement :meth:`__call__` on the subclass.
     """
 
-    async def __call__(self, in_q, out_q):
+    def __init__(self):
+        self._in_q = None
+        self._out_q = None
```
Comment (mdellweg) [inline pulpcore/plugin/stages/api.py]: I do not know if this is necessary. I could just leave it as is and if someone uses the stage without adhering to the design the error would change from something like "None has no attribute get" to "self has no attribute _in_q".

File: pulpcore/plugin/stages/api.py [hunk 2]
```diff
@@ -13,90 +13,144 @@ class Stage:
...
+        Args:
+            item: A handeld instance of :class:`pulpcore.plugin.stages.DeclarativeContent`
+        """
+        await self._out_q.put(item)
+
+    async def finish(self):
+        """
+        Coroutine to signal the next stage that this stage finished it's work.
+        """
+        await self._out_q.put(None)
```
Comment (mdellweg) [inline pulpcore/plugin/stages/api.py]: should i clean _in_q and _out_q here to make sure the stage instance does not get called a second time?

File: pulpcore/plugin/stages/api.py [hunk 3]
```diff
@@ -13,90 +13,144 @@ class Stage:
...
         raise NotImplementedError(_('A plugin writer must implement this method'))
 
-    @staticmethod
-    async def batches(in_q, minsize=50):
+    async def items(self):
         """
-        Asynchronous iterator yielding batches of :class:`DeclarativeContent` from `in_q`.
+        Asynchronous iterator yielding items of :class:`DeclarativeContent` from `self._in_q`.
+
+        The iterator will try to get as many instances of
```
Comment (dralley) [inline pulpcore/plugin/stages/api.py]: Fix docstring here.

**Comment by mdellweg** [comment]
OK, now i think, i have it. Turns out that after catching exceptions from futures, in a gather, you need to cancel all other pending futures by hand.

File: pulpcore/plugin/stages/api.py [hunk 4]
```diff
@@ -13,90 +13,148 @@ class Stage:
...
+            nonlocal batch
+            nonlocal shutdown
+            nonlocal no_block
             if content is None:
-                return True
-            batch.append(content)
-            return False
+                shutdown = True
+            else:
+                if getattr(content, "future", None):
```
Comment (gmbnomis) [inline pulpcore/plugin/stages/api.py]: This implements https://pulp.plan.io/issues/4296, right? I think this does not belong in the PR, but in a separate one. (But at least in a separate commit)

**Comment by bmbouter** [comment]
I like the Stages API improvements, like the finish() and put() methods, and abstracting away using the queues directly. I share @gmbnomis 's concerns about the `ArtifactDownloader` changes. Are we able to separate those changes into another PR? +1 to the changed gmbnomis requested.

Also can a 'Refactor' type be filed in Redmine and the commit associated with that please?

@mdellweg I really appreciate you making the Stages API better. It's really turning into something great w/ all of your and @gmbnomis contributions!

File: pulpcore/plugin/stages/artifact_stages.py [hunk 5]
```diff
@@ -218,77 +72,110 @@ class ArtifactDownloader(Stage):
...
     downloads completed. Since it's a stream the total count isn't known until it's finished.
 
-    This stage drains all available items from `in_q` and starts as many downloaders as possible
+    This stage drains all available items from `_in_q` and starts as many downloaders as possible
     (up to `download_concurrency` set on a Remote)
 
     Args:
-        max_concurrent_content (int): The maximum number of
-            :class:`~pulpcore.plugin.stages.DeclarativeContent` instances to handle simultaneously.
+        max_concurrent_downloads (int): The maximum number of
```
[the remark this answers is not in the document]
Comment (mdellweg) [reply pulpcore/plugin/stages/artifact_stages.py]: See, i never really understood that. By looking at the mentioned PR, i think it was more obvious with the other variable `max_concurrent_downloads` nearby.
Need to think this over again.

[the remark this answers is not in the document]
Comment (gmbnomis) [reply pulpcore/plugin/stages/artifact_stages.py]: Yes, looking at it, it wasn't the best explanation of `max_concurrent_content` to begin with and after  https://github.com/pulp/pulp/pull/3696 the `ArtifactDownloader` does not say much about the actual downloading. Probably, the stage's documentation should be updated to make this clearer.

File: pulpcore/plugin/stages/api.py (hunk 2 above)
Comment (mdellweg) [reply pulpcore/plugin/stages/api.py]: Sounds like a good idea.

**Comment by mdellweg** [comment]
I reduced the refactoring of the ArtifactDownload stage (folded the state into the `__call__` method anyway). And removed the non_blocking behaviour of content with future (as that really should be another PR).
The question came up, whether `finish` is the right name, or whether we should call it `shutdown`. Thoughts?

**Comment by bmbouter** [comment]
+1 to `shutdown` but I did suggest this change on irc originally. More input from others would be good.

File: pulpcore/plugin/stages/artifact_stages.py [hunk 6]
```diff
@@ -218,77 +72,110 @@ class ArtifactDownloader(Stage):
...
+        await self.put(d_content)
+
+    async def _handle_artifact(self, d_artifact, pb):
+        try:
+            try:
+                await d_artifact.download()
+            except Exception:
+                # We need to make this information available before releasing the semaphore…
+                self._unhandled_exception = True
+                await self._in_q.put(None)  # … and thawing the await on the input queue
```
[the remark this answers is not in the document]
Comment (mdellweg) [reply pulpcore/plugin/stages/artifact_stages.py]: Agree in the `not so nice`.

**Comment by mdellweg** [comment]
> One additional idea: Should/could we make use of context managers to avoid the final `.finish()`? (terminating None is send on exiting a context)

Generally, i like the idea of the Context manager. But, what should the use look like?
```
async def __call__(self):
    with self:
        pass
```
is the only thing that comes to my mind. To avoid that, i would prefer wrapping a call to `async def run(self)` in the `__call__` method followed by `self.put(None)`. And the `run` is what derived classes override, then.

Another thing, i was thinking of was making the `Stage` (and a `BatchedStage`) itself an async iterator (instead of `batches` and `items`). This could result in the use to be
```
async for item_or_batch in self:
    pass
```

**Comment by mdellweg** [comment]
What do you mean by 'returns itself'? `with open('filename') as f:` returns an open file handle (something new).
I see, from a users perspective it is not obvious, that i shall use a context manager or that `finish` call at all. Therefore i tend to go with the `run` method and leave only the instructions

  1. Override the `run` function.
  2. Take items from either sequence: `items` or `batches`
  3. Pass items with `put` to the next stage.

to the plugin writer.

**Comment by gmbnomis** [comment]
> Therefore i tend to go with the `run` method and leave only the instructions
> 
> 1. Override the `run` function.
> 2. Take items from either sequence: `items` or `batches`
> 3. Pass items with `put` to the next stage.
> 
> to the plugin writer.

@mdellweg Yes, I agree. It does the job and is much more obvious than a context manager.

**Comment by mdellweg** [comment]
Let's wait for another opinion, then i am happy to implement, as it's rather straight forward.
@bmbouter @dralley @who_else_could_be_interested?

File: pulpcore/plugin/stages/api.py [hunk 7]
```diff
@@ -3,100 +3,155 @@
...
 
-    async def __call__(self, in_q, out_q):
+    def __init__(self):
+        self._in_q = None
+        self._out_q = None
+
+    def _connect(self, next_stage, maxsize=100, profile_num=None):
         """
-        The coroutine that is run as part of this stage.
+        Create a :class:`asyncio.Queue` and connect this to the next stage.
```
Comment (jortel) [inline pulpcore/plugin/stages/api.py]: Please docstring args.

File: pulpcore/plugin/stages/api.py [hunk 8]
```diff
@@ -3,100 +3,155 @@
...
 
-        Args:
-            in_q (:class:`asyncio.Queue`): The queue to receive items from the previous stage.
-            out_q (:class:`asyncio.Queue`): The queue to put handled items into for the next stage.
+        It calls :meth:`run` and signals the next stage that its work is finished.
+        """
+        await self.run()
+        await self._out_q.put(None)
+
+    async def run(self):
```
Comment (jortel) [inline pulpcore/plugin/stages/api.py]: I assume that `run()` is not part of the _public_ `Stage` interface and meant to be overridden by subclasses so should be _protected_ `_run()`

Comment (mdellweg) [reply pulpcore/plugin/stages/api.py]: I thought, it is part of the external API in that it must be overridden by subclasses. But it should not be called from the outside.
I can rename it.

File: pulpcore/plugin/stages/api.py [hunk 9]
```diff
@@ -3,100 +3,155 @@
...
 
         Returns:
             The coroutine that runs this stage.
 
         """
         raise NotImplementedError(_('A plugin writer must implement this method'))
 
-    @staticmethod
-    async def batches(in_q, minsize=50):
+    async def items(self):
```
Comment (jortel) [inline pulpcore/plugin/stages/api.py]: Not part of the _public_ `Stage` interface.  Should be `_items()`.

File: pulpcore/plugin/stages/api.py [hunk 10]
```diff
@@ -3,100 +3,155 @@
...
+        """
+        This coroutine makes the stage callable.
 
-        Args:
-            in_q (:class:`asyncio.Queue`): The queue to receive items from the previous stage.
-            out_q (:class:`asyncio.Queue`): The queue to put handled items into for the next stage.
+        It calls :meth:`run` and signals the next stage that its work is finished.
+        """
+        await self.run()
+        await self._out_q.put(None)
```
Comment (jortel) [inline pulpcore/plugin/stages/api.py]: This is a really good change.  I unintentionally had my stage add too many _None_ to the out_q and it reeked silent havoc.  Took a while to figure out.

File: pulpcore/plugin/stages/api.py (hunk 9 above)
Comment (mdellweg) [reply pulpcore/plugin/stages/api.py]: This _should_ be called by the subclass. So i'd say, it _is_ public api. Maybe i misunderstood the reasoning behind _ing identifiers.

File: pulpcore/plugin/stages/api.py [hunk 11]
```diff
@@ -3,100 +3,155 @@
...
+                            await self.put(d_content)
+
+        """
+        while True:
+            content = await self._in_q.get()
+            if content is None:
+                break
+            yield content
+
+    async def batches(self, minsize=50):
```
Comment (jortel) [inline pulpcore/plugin/stages/api.py]: Not part of the _public_ `Stage` interface.  Should be `_batches()`.

File: pulpcore/plugin/stages/api.py [hunk 12]
```diff
@@ -3,100 +3,155 @@
...
                     break
                 else:
-                    shutdown = add_to_batch(batch, content)
+                    _add_to_batch(content)
 
             if batch and (len(batch) >= minsize or shutdown):
                 yield batch
                 batch = []
 
+    async def put(self, item):
```
Comment (jortel) [inline pulpcore/plugin/stages/api.py]: Not part of the _public_ `Stage` interface.  Should be `_put()`.

File: pulpcore/plugin/stages/content_unit_stages.py [hunk 13]
```diff
@@ -65,8 +59,7 @@ class QueryExistingContentUnits(Stage):
                             continue
                         declarative_content.content = result
             for declarative_content in batch:
-                await out_q.put(declarative_content)
-        await out_q.put(None)
+                await self.put(declarative_content)
 
 
 class ContentUnitSaver(Stage):
```
Comment (jortel) [inline pulpcore/plugin/stages/content_unit_stages.py]: While we're refactoring ... there is no `ContentUnit` class in pulp3.  Would prefer dropping `Unit` from the name to just be `ContentSaver` but not a big deal.

File: pulpcore/plugin/stages/api.py [hunk 14]
```diff
@@ -3,100 +3,155 @@
...
-    To make a stage, inherit from this class and implement :meth:`__call__` on the subclass.
+    To make a stage, inherit from this class and implement :meth:`run` on the subclass.
     """
 
-    async def __call__(self, in_q, out_q):
+    def __init__(self):
+        self._in_q = None
+        self._out_q = None
+
+    def _connect(self, next_stage, maxsize=100, profile_num=None):
```
Comment (jortel) [inline pulpcore/plugin/stages/api.py]: :+1:  This is appropriate behavor for the stage.

This needs to be part of the _public_ interface so should be `connect()`.

File: pulpcore/plugin/stages/content_unit_stages.py (hunk 13 above)
Comment (mdellweg) [reply pulpcore/plugin/stages/content_unit_stages.py]: Tried to find all the *Unit* occurrences.

**Comment by mdellweg** [comment]
@jortel ready for rereview

**Comment by jortel** [comment]
@mdellweg: 
Thank you for making the requested changes. After further consideration, reading of python docs[1], and discussions, it seems that python does not actually support _protected_ access.  The intent of the `_` prefix is to denote _private_ implementation details with no guarantee of stability even to subclasses.  With this in mind, even methods used/overridden by subclasses should be _public_.  As as result, I'm going to ask you to basically revert the changes I requested regarding _protected_ methods.  In addition to this, make an additional change re: `Stage.connect()`.

In summary:

1. All of the `Stage` methods and attributes will be _public_ (no `_` prefix).  This includes the `in_q` and `out_q` attributes.
2. Get rid of `Stage._connect()` and move the logic to `create_pipeline()` which will construct the queues and set them directly on `Stage.in_q=` and `Stage.out_q=`.

Sorry for the faulty guidance and extra work.  With these changes, I feel confident that we can get this merged very soon.

[1] https://docs.python.org/3.6/tutorial/classes.html#tut-private

**Comment by bmbouter** [comment]
I also learned about Python more by reading these docs. I agree w/ reverting to the public names and making the change to have `create_pipeline()` handle the Queue creation along with pipeline construction.

**Comment by gmbnomis** [comment]
> In summary:
> 
> 1. All of the `Stage` methods and attributes will be _public_ (no `_` prefix).  This includes the `in_q` and `out_q` attributes.

@jortel l I fully agree to the methods part. But I don't agree on `in_q` and `out_q`. My impression was that the whole point of this refactoring was to hide the underlying implementation based on queues. For example, the `ArtifactDownloader` stage could be simpler when using `in_q` direclty, but @mdellweg changed it to use the async iterator protocol (which was the correct decision).

IMHO, `_in_q` and `_out_q` are to be considered private indeed.

> 2. Get rid of `Stage._connect()` and move the logic to `create_pipeline()` which will construct the queues and set them directly on `Stage.in_q=` and `Stage.out_q=`.  The reasoning behind this is that building the pipeline is the sole concern of `create_pipeline()` and not the `Stage`.

I think that the check whether a stage is reused still belongs into the stage. Thus, I propose to still have a `connect(self, in_q, out_q)`.  Creating the queues outside (including the profiling distinction) is a good idea.

**Comment by mdellweg** [comment]
I can fully support, the last post from @gmbnomis , and will partially revert the reverting...
This should give us the most consistent version.

**Comment by jortel** [comment]
I'm fine with the `_in_q` and `_out_q` attributes being _private_ and set by a _public_ `connect()` method as proposed by @gmbnomis.

Any objection @bmbouter?

**Comment by mdellweg** [comment]
I should mention, i tried to streamline the use of the variable names `d_artifact` and `d_content` because they are half way between the non-descriptive `da`, `dc` and way less clunky than 'declarative_something`.

File: pulpcore/plugin/stages/api.py [hunk 15]
```diff
@@ -10,93 +10,136 @@ class Stage:
...
-    To make a stage, inherit from this class and implement :meth:`__call__` on the subclass.
+    To make a stage, inherit from this class and implement :meth:`run` on the subclass.
     """
 
-    async def __call__(self, in_q, out_q):
+    def __init__(self):
+        self._in_q = None
+        self._out_q = None
+
+    def connect(self, in_q, out_q):
```
Comment (jortel) [inline pulpcore/plugin/stages/api.py]: missing docstring.

File: pulpcore/plugin/stages/api.py [hunk 16]
```diff
@@ -10,93 +10,136 @@ class Stage:
...
-                                await out_q.put(declarative_content)
-                        await out_q.put(None)
+                                await self.put(d_content)
 
         """
         batch = []
         shutdown = False
 
-        def add_to_batch(batch, content):
+        def _add_to_batch(content):
```
Comment (jortel) [inline pulpcore/plugin/stages/api.py]: `_` prefix on closure not needed.

**Comment by mdellweg** [comment]
Let's hope, this version really meets all requirements.

A word for explanation: As discussed with @bmbouter , '_ing' identifiers is just a convention to mark them as not being part of any stable api. This is to be understood like in C++, where a **friend** class or method has access to **private** members, and must therefore be considered as an integral part of the class at hand. Here it means, that `Stage` and `create_pipeline` are supposed to present a common API at the module boundary (here `pulpcore.plugin`). The question whether anything should be protected, should be the same as should it be visible to the plugin writer, and should it be mentioned in the api-docs.

**Comment by gmbnomis** [comment]
@bmbouter @mdellweg Thinking about this from a plugin API perspective makes perfect sense to me. And, yes, I agree that things like `stages[i]._connect(...)` are usual in Python. Or, as GvR put it (allegedly): "We're all adults here."

**Comment by jortel** [comment]
The encapsulation principal of _object-oriented_ programming is clear.  The boundary is the **object** not the module or the package or the project.  The value of encapsulation is to insulate the user of an object against changes in implementation details of either methods or object state (attributes).  Further, it protects the object's internal integrity.  Encapsulation is provided in two ways.  1. The black-box nature of how a method is implemented.  That is, a method signature & behavior is guaranteed but the implementation  details are not.  2. _Private_ members are part of an object's internal implementation and must never be accessed.  True OO languages actually enforce this for good reason.

This brings us to python and it's definition of _private_.  Although, it only characterizes that _private_ methods are _"an implementation detail"_ (of the object) and _"subject to change without notice"_.  Neither of those statements lead me to believe that _private_ members should ever be accessed outside the class.  Doing so is unsafe for the object and the code using the object for all the reasons outlined by the _object-oriented_ encapsulation principal.  I don't think that omission of a statement like _"private members should not be accesses externally"_ implies that basic OO encapsulation principals don't apply.  Rather, IMHO, it's assumed.

As for C++ _friend_ classes - the relationship explicitly redefines the boundary of encapsulation in a between two specific classes.  This is different than ignoring the boundary all together. For C++ projects I worked on (almost 10 years), _friends_ were widely regarded as an anti-pattern and highly discouraged or prohibited. 

That said, _python_ does not enforce encapsulation at the object boundary (or any boundary for that matter) so it's reasonable to assume that violating it in an informed, deliberate way is condoned.  There seems to be consensus to make `Stage._connect()` private and called by `create_pipeline()`.  I will not oppose it further.

**Comment by dralley** [comment]
My PR seems to have caused a merge conflict, sorry about that.  It should be easy to resolve since I only changed indentation w/r/t overlap.

**Comment by mdellweg** [comment]
Is the privacy thing a discussion that still needs mediation?
I think we can change that later without breaking API again.

@jortel @bmbouter

**Comment by jortel** [comment]
> Is the privacy thing a discussion that still needs mediation?

No.

> I think we can change that later without breaking API again.

Probably.

> 
> @jortel @bmbouter

Just need @bmbouter to (re)review, approve and merge.  He'll be back on Monday.

**Comment by mdellweg** [comment]
I guess, having his approval on irc private message is just not enough... ;(

File: pulpcore/tests/unit/stages/test_artifactdownloader.py [hunk 17]
```diff
@@ -49,6 +61,8 @@ def setUp(self):
         super().setUp()
         DownloaderMock.reset()
         self.now = 0
+        self.in_q = asyncio.Queue()
+        self.out_q = asyncio.Queue()
```
Comment (dralley) [inline pulpcore/tests/unit/stages/test_artifactdownloader.py]: Why does the mock downloader use un-prefixed names while the actual downloader uses prefixed names?

Sorry if it's already been discussed, I haven't kept track of the entire discussion.

Comment (mdellweg) [reply pulpcore/tests/unit/stages/test_artifactdownloader.py]: These are the mocked queues saved in the test instance. The download stage will still use the masked names internally.

File: docs/api-reference/stages.rst [hunk 18]
```diff
@@ -60,11 +60,11 @@ Artifact Related Stages
 Content Related Stages
 ^^^^^^^^^^^^^^^^^^^^^^
 
-.. autoclass:: pulpcore.plugin.stages.ContentUnitSaver
+.. autoclass:: pulpcore.plugin.stages.ContentSaver
    :special-members: __call__
    :private-members: _pre_save, _post_save
 
-.. autoclass:: pulpcore.plugin.stages.QueryExistingContentUnits
+.. autoclass:: pulpcore.plugin.stages.QueryExistingContents
```
Comment (dralley) [inline docs/api-reference/stages.rst]: Would anyone mind if this were renamed to ```QueryExistingContent```?  The extra 's' can trip you up a bit, since it's the only one named like that.

Comment (dralley) [reply docs/api-reference/stages.rst]: @jortel @bmbouter @mdellweg

**Comment by mdellweg** [comment]
Thanks to all for the vivid discussion!
Thanks @bmbouter for the timing. We are just in time for the PulpPluginDay. Greetings from CfgMgmtCamp!
